File: solutions/typescript/0075.ts
/**
 * 75. 颜色分类
 * Do not return anything, modify nums in-place instead.
 */
function sortColors(nums: number[]): void {
    let low = 0;
    let current = 0;
    let high = nums.length - 1;

    while (current <= high) {
        if (nums[current] === 0) {
            [nums[low], nums[current]] = [nums[current], nums[low]];
            low++;
            current++;
        } else if (nums[current] === 2) {
            [nums[current], nums[high]] = [nums[high], nums[current]];
            high--;
        } else {
            current++;
        }
    }
}

export default sortColors;

File: solutions/rust/0075.rs
/// 75. 颜色分类
/// 荷兰国旗算法，三指针一趟扫描原地排序
pub fn sort_colors(nums: &mut Vec<i32>) {
    let mut low = 0usize;
    let mut current = 0usize;
    let mut high = nums.len();

    while current < high {
        match nums[current] {
            0 => {
                nums.swap(low, current);
                low += 1;
                current += 1;
            }
            2 => {
                high -= 1;
                nums.swap(current, high);
            }
            _ => {
                current += 1;
            }
        }
    }
}

impl Solution {
    pub fn sort_colors(nums: &mut Vec<i32>) {
        sort_colors(nums);
    }
}
